opal/mca/accelerator/cuda/accelerator_cuda.c

```diff
@@ -0,0 +1,690 @@
+    CUresult result;
+
+    result = opal_accelerator_cuda_func.cuMemAlloc(&tmp, size);
+    if (OPAL_UNLIKELY(CUDA_SUCCESS != result)) {
+        return OPAL_ERROR;
+    }
+    result = opal_accelerator_cuda_func.cuMemcpyAsync(tmp, (CUdeviceptr) src, size, opal_accelerator_cuda_memcpy_stream);
+    if (OPAL_UNLIKELY(CUDA_SUCCESS != result)) {
+        opal_show_help("help-accelerator-cuda.txt", "cuMemcpyAsync failed", true, tmp, src, size,
+                       result);
+        return OPAL_ERROR;
+    }
+    result = opal_accelerator_cuda_func.cuMemcpyAsync((CUdeviceptr) dest, tmp, size, opal_accelerator_cuda_memcpy_stream);
+    if (OPAL_UNLIKELY(CUDA_SUCCESS != result)) {
+        opal_show_help("help-accelerator-cuda.txt", "cuMemcpyAsync failed", true, dest, tmp,
+                       size, result);
+        return OPAL_ERROR;
+    }
+    result = opal_accelerator_cuda_func.cuStreamSynchronize(opal_accelerator_cuda_memcpy_stream);
+    if (OPAL_UNLIKELY(CUDA_SUCCESS != result)) {
+        opal_show_help("help-accelerator-cuda.txt", "cuStreamSynchronize failed", true,
+                       OPAL_PROC_MY_HOSTNAME, result);
+        return OPAL_ERROR;
+    }
+    opal_accelerator_cuda_func.cuMemFree(tmp);
+    return OPAL_SUCCESS;
+}
+
+static int accelerator_cuda_malloc(void **ptr, size_t size)
+{
+    CUresult result;
+    if (size > 0) {
+        result = opal_accelerator_cuda_func.cuMemAlloc((CUdeviceptr *) ptr, size);
+        if (OPAL_UNLIKELY(CUDA_SUCCESS != result)) {
+            opal_show_help("help-accelerator-cuda.txt", "cuMemAlloc failed", true,
+                           OPAL_PROC_MY_HOSTNAME, result);
+            return result;
+        }
+    }
+    return 0;
+}
+
+static int accelerator_cuda_free(void *ptr)
+{
+    CUresult result;
+    if (NULL != ptr) {
+        result = opal_accelerator_cuda_func.cuMemFree((CUdeviceptr) ptr);
+        if (OPAL_UNLIKELY(CUDA_SUCCESS != result)) {
+            opal_show_help("help-accelerator-cuda.txt", "cuMemFree failed", true,
+                           OPAL_PROC_MY_HOSTNAME, result);
+            return result;
+        }
+    }
+    return 0;
+}
+
+static int accelerator_cuda_get_address_range(const void *ptr, void **base,
+                                       size_t *size)
+{
+    CUresult result;
+    result = opal_accelerator_cuda_func.cuMemGetAddressRange((CUdeviceptr *) base, size, (CUdeviceptr) ptr);
+    if (OPAL_UNLIKELY(CUDA_SUCCESS != result)) {
+        opal_show_help("help-accelerator-cuda.txt", "cuMemGetAddressRange failed 2", true,
+                       OPAL_PROC_MY_HOSTNAME, result, ptr);
+        return OPAL_ERROR;
+    } else {
+        opal_output_verbose(50, opal_accelerator_base_framework.framework_output,
+                            "CUDA: cuMemGetAddressRange passed: addr=%p, pbase=%p, psize=%lu ",
+                            ptr, *(char **) base, *size);
+    }
+    return 0;
+}
+
+static bool accelerator_cuda_is_ipc_enabled(void)
+{
+    return true;
```

Comment:
How do we handle instances with one accelerator?

Reply:
I was thinking that some API's wouldn't have IPC support so this could be checked. I suppose if it's really necessary it can be added when someone doesn't support IPC

Reply:
Going to move IPC code to a separate PR